lib/EarlyFunctionAnalysis/UsedRegisters.h

```diff
@@ -0,0 +1,152 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <map>
+#include <set>
+
+#include "revng/EarlyFunctionAnalysis/CallGraph.h"
+#include "revng/EarlyFunctionAnalysis/Common.h"
+#include "revng/MFP/SetLattices.h"
+#include "revng/Model/Register.h"
+#include "revng/Support/MetaAddress.h"
+
+#include "DetectABI.h"
+
+namespace efa {
+
+struct FunctionABI {
+  using SetOfRegisters = std::set<model::Register::Values>;
+  using SUL = SetUnionLattice<SetOfRegisters>;
+  SetOfRegisters ArgumentRegisters;
+  SetOfRegisters ReturnRegisters;
+};
+
+// using LatticeElement = std::map<MetaAddress, FunctionABI>;
+
+struct LatticeElement : std::map<MetaAddress, FunctionABI> {
+  virtual ~LatticeElement() = default;
+
+  [[nodiscard]] virtual bool
+  hasRetReg(MetaAddress Addr, model::Register::Values V) const {
+    if (count(Addr)) {
+      return at(Addr).ReturnRegisters.count(V);
+    }
+
+    return false;
+  }
+
+  [[nodiscard]] virtual bool
+  hasArgReg(MetaAddress Addr, model::Register::Values V) const {
+    if (count(Addr)) {
+      return at(Addr).ArgumentRegisters.count(V);
+    }
+
+    return false;
+  }
+};
+
+struct ExtremalLatticeElement : LatticeElement {
```

Comment:
This does not need to be a class. It just need to be a specific instance of `LatticeElement`.

Reply:
You're right, but you're describing the "top" element more than the extremal value.

But in this case, I think we can have "bottom" as the extremal value.
I don't see a reason to start from "top".